packages/astro/src/vite-plugin-build-html/index.ts

```diff
@@ -0,0 +1,392 @@
+const matchSrcset: typeof srcsetParse = (srcsetParse as any).default;
+
+const PLUGIN_NAME = '@astro/rollup-plugin-build';
+const ASTRO_PAGE_PREFIX = '@astro-page';
+const ASTRO_SCRIPT_PREFIX = '@astro-script';
+
+const ASTRO_EMPTY = '@astro-empty';
+
+const tagsWithSrcSet = new Set(['img', 'source']);
+
+const isAstroInjectedLink = (node: parse5.Element) => isStylesheetLink(node) && getAttribute(node, 'data-astro-injected') === '';
+const isBuildableLink = (node: parse5.Element, srcRoot: string) => isAstroInjectedLink(node) || getAttribute(node, 'href')?.startsWith(srcRoot);
+const isBuildableImage = (node: parse5.Element, srcRoot: string) => getTagName(node) === 'img' && getAttribute(node, 'src')?.startsWith(srcRoot);
+const hasSrcSet = (node: parse5.Element) => tagsWithSrcSet.has(getTagName(node)) && !!getAttribute(node, 'srcset');
+
+interface PluginOptions {
+  astroConfig: AstroConfig;
+  astroStyleMap: Map<string, string>;
+  astroPageStyleMap: Map<string, string>;
+  chunkToReferenceIdMap: Map<string, string>;
+  logging: LogOptions;
+  allPages: AllPagesData;
+  pageNames: string[];
+  pureCSSChunks: Set<RenderedChunk>;
+  origin: string;
+  routeCache: RouteCache;
+  viteServer: ViteDevServer;
+}
+
+export function rollupPluginAstroBuildHTML(options: PluginOptions): VitePlugin {
+  const { astroConfig, astroStyleMap, astroPageStyleMap, chunkToReferenceIdMap, pureCSSChunks, logging, origin, allPages, routeCache, viteServer, pageNames } = options;
+
+  const srcRoot = astroConfig.src.pathname;
+
+  // A map of pages to rendered HTML
+  const renderedPageMap = new Map<string, string>();
+
+  //
+  const astroScriptMap = new Map<string, string>();
+  const astroPageMap = new Map<string, string>();
+  const astroAssetMap = new Map<string, Promise<Buffer>>();
+
+  const cssChunkMap = new Map<string, string[]>();
+
+  return {
+    name: PLUGIN_NAME,
+
+    enforce: 'pre',
+
+    async options(inputOptions) {
+      const htmlInput: Set<string> = new Set();
+      const assetInput: Set<string> = new Set(); // TODO remove?
+      const jsInput: Set<string> = new Set();
+
+      for(const [component, pageData] of Object.entries(allPages)) {
+        const [renderers, mod] = pageData.preload;
+
+        for(const path of mod.$$metadata.getAllHydratedComponentPaths()) {
+          jsInput.add(path);
+        }
+
+        for(const pathname of pageData.paths) {
+          pageNames.push(pathname.replace(/\/?$/, '/index.html').replace(/^\//, ''));
+          const id = ASTRO_PAGE_PREFIX + pathname;
+          const html = await ssrRender(renderers, mod, {
+            astroConfig,
+            filePath: new URL(`./${component}`, astroConfig.projectRoot),
+            logging,
+            mode: 'production',
+            origin,
+            pathname,
+            route: pageData.route,
+            routeCache,
+            viteServer,
+          });
+          renderedPageMap.set(id, html);
+
+          const document = parse5.parse(html, {
```

Comment:
Just a sidenote: if we like parse5 better we should probably replace htmlparser2 with this (I didn’t profile parse5; I just went with htmlparser2 because it’s fast, easy-to-use, and actively maintained)

Reply:
Was going to be my comment as well. I'm fine with either, but we should only use one.